### DISCUSSION TITLE
my app started to appear too much anr about react native firebase

### DISCUSSION THREAD

**Discussion by emrecanozkok** [invertase/react-native-firebase#5680] [Q&A]
im facing a anr problem about 4 months and i cant fix it.

the error is

`Broadcast of Intent { act=com.google.android.c2dm.intent.RECEIVE flg=0x11000010 pkg=com.xxxyyy cmp=com.xxxyyy/io.invertase.firebase.messaging.ReactNativeFirebaseMessagingReceiver (has extras) }`

![image](https://user-images.githubusercontent.com/8539564/131855845-0e610433-5502-48d5-8b56-d8e03d442267.png)

app working nice but some devices facing problem.actually i dont have any idea to fix it.and i cant find any solution.

![image](https://user-images.githubusercontent.com/8539564/131856603-b6d4ea4b-f037-476d-9428-ffd1e2f6dd02.png)

![image](https://user-images.githubusercontent.com/8539564/131856270-cd22527c-cd65-45d1-9d8b-e2c5d6a4daff.png)
![image](https://user-images.githubusercontent.com/8539564/131856315-11f5787d-b8eb-4e8f-a880-b0529ca46f6d.png)
![image](https://user-images.githubusercontent.com/8539564/131856349-032b209f-b40b-43c4-a41e-5c43b3ac22d1.png)

can someone help me, thank you.

**Comment by mikehardy** [comment]
So, ANRs are tough to fix. They are a special form of performance optimization, and if you've never done performance optimization the first thing you need to know is that you must 

a) have a good reproduction - you may need to purchase some low-end android phone that reproduces the problem and learn how to get it set up so you see the problem
b) have a good set of tools and a workbench to generate actionable information - for ANRs this usually means enabling StrictMode in android and carefully watching the log
c) a really good change/test/report cycle where you fix items one by one until you have enough of them fixed

I believe react-native-firebase is *implicated* here but I do not believe it is at fault

What do I mean by that? Well, the firebase broadcast receiver is the *entry point* for the app to be woken up, by a remote notification. So, this module shows up in your reports, because we are the handle used to begin app execution.

However, the app is (for whatever reason) taking longer than 10 seconds to boot and then respond the remote notification. Why? Who knows! You could literally program a javascript hook that sleeps for 11 seconds, you would get an ANR and it would look like react-native-firebase receiver triggered the ANR. But that is not the module's fault is it?

There is more good information here, to help you on the journey of app startup performance optimization / ANR avoidance:
https://stackoverflow.com/a/49896457/9910298

**Reply by emrecanozkok** [reply]
As I understand it, the problem is the overall performance of the application? Or is it the opening performance? because my application is not complex and it is an application that only has 2 pages and makes crud requests to 2 APIs.

**Reply by mikehardy** [reply]
I have learned over decades of programming to take statements like this:

> because my application is not complex and it is an application that only has 2 pages and makes crud requests to 2 APIs.

and ignore them, with no offense intended at all! I can't tell you how many times performance testing on my code has surprised me with unexpectedly bad results in the simplest code.

You have to test it.

It is related to app startup performance, the Android system receives an FCM, it calls the react-native-firebase receiver, your app starts to boot, and doesn't boot in time, and you get an ANR.

You have to measure app startup performance starting from an FCM to completely handling the FCM and improve where ever it is slow.

**Reply by emrecanozkok** [reply]
I can't say the opposite of what you said. I am mainly a php programmer and it takes some time to understand the logic of the mobile application. I just want to ask 2 questions and I would like to close this topic.

Even if we don't click on the notification, is there a slowness about the app when it first arrives?
Does this slowness occur when we click on the notification?

thank you.

**Reply by mikehardy** [reply]
If you send FCM to the app and it contains no notification block, just a data block, your app will be booted and do it's startup when the message arrives, with no user interaction at all.

If you send FCM with a notification block it will post the notification but I do not think it will boot the app, at that point the app will not boot until the user interacts with the notification in any way other than to just dismiss it.

That's how it is supposed to work I think, but as you perform testing on it your experimental results will of course take precedence

**Reply by emrecanozkok** [reply]
I bought the device that gave the most errors. This device is samsung a21s. I updated to android 11 api30 level, which is the version that gives error. I'm looking into it in detail but it doesn't produce any anr. there are no problems with notifications. I send in Data block in every notification because if the application is open I update some parts. I don't think this situation produces anr because this problem occurs in a very limited device list. If something like you said had occurred, I hope it would have occurred more on s4 and s5. I respect their opinion, but I don't think that's the reason for the problem. I think this might be a problem with the anr detection system on google side. Because android studio does not give any warning. All logs are info or react native logs that I printed.

**Reply by mikehardy** [reply]
Perhaps start a few apps at a time on the device, maybe make a quick "useless" app that does nothing but runs and uses CPU so it slows the device down in general, then try your app. I've never heard of a problem with "the ANR detection system" on devices. It's just a timer that is making sure the main thread isn't busy for too long in wall-clock time, and it's quite reliable.

**Reply by emrecanozkok** [reply]
Thanks for your valuable comments.
Another thing I'm curious about; Why does this anr not appear on slow devices and only on android 11 devices? What could have changed?

**Reply by mikehardy** [reply]
Unsure, sorry
